File: compilers/java/src/test/java/runar/compiler/passes/BranchLiftThenArmTest.java
package runar.compiler.passes;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertInstanceOf;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Test;
import runar.compiler.frontend.ParserDispatch;
import runar.compiler.ir.anf.AnfBinding;
import runar.compiler.ir.anf.AnfMethod;
import runar.compiler.ir.anf.AnfProgram;
import runar.compiler.ir.anf.AnfValue;
import runar.compiler.ir.anf.BytesConst;
import runar.compiler.ir.anf.If;
import runar.compiler.ir.anf.LoadConst;
import runar.compiler.ir.anf.LoadProp;
import runar.compiler.ir.anf.UpdateProp;
import runar.compiler.ir.ast.ContractNode;

/**
 * Regression test: the branch-lift must not zero the matched arm.
 *
 * <p>{@code AnfLower.liftBranchUpdateProps} flattens a dispatch chain
 *
 * <pre>
 *   if (p == 0n) { this.c0 = v; } else if (p == 1n) { this.c1 = v; }
 *   else { assert(false); }
 * </pre>
 *
 * into one single-valued {@code if} per property plus a top-level
 * {@code update_prop}. The {@code if}'s then-arm must evaluate to the assigned
 * value and its else-arm to the property's old value.
 *
 * <p>The defect: the then-arm was built from {@code branch.valueBindings} —
 * everything BEFORE the {@code update_prop} in the original arm. That ends on
 * the assigned value only when the value was computed INSIDE the arm. When the
 * arm assigns something bound outside it, {@code valueBindings} is empty, the
 * arm was emitted EMPTY, and stack lowering padded it with a zero push
 * ({@code OP_IF OP_0 OP_ELSE OP_DUP OP_ENDIF}): the MATCHED branch wrote 0.
 *
 * <p>{@code examples/ts/tic-tac-toe} escapes it only because
 * {@code this.cN = this.turn} puts a {@code load_prop} inside the arm — that
 * shape is the control below.
 */
class BranchLiftThenArmTest {

    private static final String LOCAL_VALUE_DISPATCH = """
        class LocalValueDispatch extends StatefulSmartContract {
          c0: bigint;
          c1: bigint;

          constructor(c0: bigint, c1: bigint) {
            super(c0, c1);
            this.c0 = c0;
            this.c1 = c1;
          }

          public poke(position: bigint, value: bigint) {
            const doubled: bigint = value + value;
            if (position == 0n) { this.c0 = doubled; }
            else if (position == 1n) { this.c1 = doubled; }
            else { assert(false); }
          }
        }
        """;

    private static final String IN_ARM_VALUE_DISPATCH = """
        class InArmValueDispatch extends StatefulSmartContract {
          c0: bigint;
          c1: bigint;
          turn: bigint;

          constructor(c0: bigint, c1: bigint, turn: bigint) {
            super(c0, c1, turn);
            this.c0 = c0;
            this.c1 = c1;
            this.turn = turn;
          }

          public poke(position: bigint) {
            if (position == 0n) { this.c0 = this.turn; }
            else if (position == 1n) { this.c1 = this.turn; }
            else { assert(false); }
          }
        }
        """;

    private record Lifted(String prop, List<AnfBinding> then, List<AnfBinding> elseArm) {}

    /**
     * Every top-level {@code update_prop} whose value is an {@code if} binding,
     * paired with that {@code if}'s two arms.
     */
    private static List<Lifted> liftedAssignments(String source, String file) throws Exception {
        ContractNode contract = ParserDispatch.parse(source, file);
        AnfProgram prog = AnfLower.run(contract);

        AnfMethod method = null;
        for (AnfMethod m : prog.methods()) {
            if (m.name().equals("poke")) {
                method = m;
                break;
            }
        }
        assertTrue(method != null, "method poke not found in lowered program");

        Map<String, AnfValue> byName = new HashMap<>();
        for (AnfBinding b : method.body()) byName.put(b.name(), b.value());

        List<Lifted> out = new ArrayList<>();
        for (AnfBinding b : method.body()) {
            if (!(b.value() instanceof UpdateProp up)) continue;
            AnfValue producer = byName.get(up.value());
            if (!(producer instanceof If iff)) continue;
            out.add(new Lifted(up.name(), iff.thenBranch(), iff.elseBranch()));
        }
        return out;
    }

    @Test
    void thenArmCarriesValueBoundOutsideTheArm() throws Exception {
        List<Lifted> lifted = liftedAssignments(LOCAL_VALUE_DISPATCH, "LocalValueDispatch.runar.ts");

        // Both properties in the chain must be lifted. If this is 0 the pass has
        // stopped recognising the shape and the arm assertions below would pass
        // vacuously.
        assertEquals(2, lifted.size(), "expected 2 lifted conditional assignments");

        for (Lifted l : lifted) {
            assertFalse(l.then().isEmpty(),
                "then-arm for this." + l.prop() + " is empty; stack lowering pads it with OP_0, "
                + "so the MATCHED branch writes zero instead of the assigned value");
            AnfBinding last = l.then().get(l.then().size() - 1);
            LoadConst lc = assertInstanceOf(LoadConst.class, last.value(),
                "then-arm for this." + l.prop() + " must end on the assigned local");
            BytesConst bc = assertInstanceOf(BytesConst.class, lc.value());
            assertEquals("@ref:doubled", bc.hex(),
                "then-arm for this." + l.prop() + " must end on the assigned local");
            assertFalse(l.elseArm().isEmpty(), "else-arm for this." + l.prop() + " is empty");
        }
    }

    /**
     * Control: the TicTacToe shape already computed its value inside the arm and
     * was always correct. The fix must add nothing here — a second binding would
     * move the checked-in goldens.
     */
    @Test
    void inArmValueShapeIsUnchanged() throws Exception {
        List<Lifted> lifted = liftedAssignments(IN_ARM_VALUE_DISPATCH, "InArmValueDispatch.runar.ts");

        assertEquals(2, lifted.size(), "expected 2 lifted conditional assignments");
        for (Lifted l : lifted) {
            assertEquals(1, l.then().size(),
                "then-arm for this." + l.prop() + " should hold exactly the in-arm load_prop");
            LoadProp lp = assertInstanceOf(LoadProp.class, l.then().get(0).value());
            assertEquals("turn", lp.name());
        }
    }
}

File: packages/runar-compiler/src/__tests__/branch-lift-then-arm.test.ts
/**
 * Regression test: the branch-lift must not zero the matched arm.
 *
 * `liftBranchUpdateProps` (04-anf-lower.ts) flattens a dispatch chain
 *
 *     if (p == 0n) { this.c0 = v; } else if (p == 1n) { this.c1 = v; }
 *     else { assert(false); }
 *
 * into one single-valued `if` per property plus a top-level `update_prop`.
 * The `if`'s then-arm is supposed to evaluate to the assigned value and its
 * else-arm to the property's old value.
 *
 * The defect: the then-arm was built from `branch.valueBindings` — everything
 * BEFORE the `update_prop` in the original arm. That only happens to end on the
 * assigned value when the value was computed INSIDE the arm. When the arm
 * assigns something bound outside it (a local, a parameter, anything hoisted
 * before the chain) `valueBindings` is empty, so the arm was emitted EMPTY and
 * stack lowering padded it with OP_0:
 *
 *     OP_IF OP_0 OP_ELSE OP_DUP OP_ENDIF      (63 00 67 76 68)
 *
 * i.e. the MATCHED branch pushed zero. `if (p == 0n) { this.c0 = local; }`
 * compiled to `this.c0 = 0` — a silent state-corrupting miscompile in every
 * tier, since all seven run the same pass.
 *
 * `examples/ts/tic-tac-toe/TicTacToe.runar.ts` escapes it only because it
 * writes `this.cN = this.turn`, whose `load_prop` lands inside the arm.
 *
 * These tests assert on BOTH levels — the ANF (then-arm non-empty and ending
 * on the assigned value) and the emitted bytes — because either alone can pass
 * for the wrong reason. The control locks in that an arm which LEGITIMATELY
 * assigns zero still emits the zero push, so the fix cannot be satisfied by
 * suppressing the byte pattern.
 */
import { describe, it, expect } from 'vitest';
import { compile } from '../index.js';

/** OP_IF OP_0 OP_ELSE OP_DUP OP_ENDIF — the matched branch pushing zero. */
const ZEROED_ARM = '6300677668';

/** Count byte-aligned occurrences of a hex sequence. */
function countByteSequence(hex: string, seq: string): number {
  let count = 0;
  for (let i = 0; i + seq.length <= hex.length; i += 2) {
    if (hex.slice(i, i + seq.length) === seq) count++;
  }
  return count;
}

/**
 * The dispatch chain assigns a LOCAL bound before the chain, so nothing in the
 * arm computes the value. This is the shape that miscompiled.
 */
const LOCAL_VALUE_DISPATCH = `
class LocalValueDispatch extends StatefulSmartContract {
  c0: bigint;
  c1: bigint;

  constructor(c0: bigint, c1: bigint) {
    super(c0, c1);
    this.c0 = c0;
    this.c1 = c1;
  }

  public poke(position: bigint, value: bigint) {
    const doubled: bigint = value + value;
    if (position == 0n) { this.c0 = doubled; }
    else if (position == 1n) { this.c1 = doubled; }
    else { assert(false); }
  }
}
`;

/**
 * Same chain assigning a PARAMETER directly — also nothing bound in the arm.
 */
const PARAM_VALUE_DISPATCH = `
class ParamValueDispatch extends StatefulSmartContract {
  c0: bigint;
  c1: bigint;

  constructor(c0: bigint, c1: bigint) {
    super(c0, c1);
    this.c0 = c0;
    this.c1 = c1;
  }

  public poke(position: bigint, value: bigint) {
    if (position == 0n) { this.c0 = value; }
    else if (position == 1n) { this.c1 = value; }
    else { assert(false); }
  }
}
`;

/**
 * CONTROL: the arms genuinely assign the literal 0. The zero push is correct
 * here and must survive — a fix that just suppresses the byte pattern fails
 * this.
 */
const LITERAL_ZERO_DISPATCH = `
class LiteralZeroDispatch extends StatefulSmartContract {
  c0: bigint;
  c1: bigint;

  constructor(c0: bigint, c1: bigint) {
    super(c0, c1);
    this.c0 = c0;
    this.c1 = c1;
  }

  public clear(position: bigint) {
    if (position == 0n) { this.c0 = 0n; }
    else if (position == 1n) { this.c1 = 0n; }
    else { assert(false); }
  }
}
`;

/**
 * CONTROL: the arms compute the value inside themselves (the TicTacToe shape).
 * Already correct before the fix; must stay correct and must not gain bytes.
 */
const IN_ARM_VALUE_DISPATCH = `
class InArmValueDispatch extends StatefulSmartContract {
  c0: bigint;
  c1: bigint;
  turn: bigint;

  constructor(c0: bigint, c1: bigint, turn: bigint) {
    super(c0, c1, turn);
    this.c0 = c0;
    this.c1 = c1;
    this.turn = turn;
  }

  public poke(position: bigint) {
    if (position == 0n) { this.c0 = this.turn; }
    else if (position == 1n) { this.c1 = this.turn; }
    else { assert(false); }
  }
}
`;

/**
 * The lifted, flattened form: every top-level `update_prop` whose value is an
 * `if` binding, paired with that `if`.
 */
function liftedAssignments(source: string) {
  const result = compile(source);
  expect(result.success).toBe(true);
  const method = result.anf!.methods.find(m => m.name === 'poke' || m.name === 'clear');
  expect(method).toBeDefined();
  const bindings = method!.body;

  const pairs: Array<{ prop: string; then: unknown[]; else: unknown[] }> = [];
  for (const b of bindings) {
    if (b.value.kind !== 'update_prop') continue;
    const valueRef = b.value.value;
    const producer = bindings.find(x => x.name === valueRef);
    if (!producer || producer.value.kind !== 'if') continue;
    pairs.push({ prop: b.value.name, then: producer.value.then, else: producer.value.else });
  }
  return { result, pairs };
}

describe('branch-lift: the matched arm carries the assigned value', () => {
  it('does not drop the then-arm when the value is a local bound before the chain', () => {
    const { pairs } = liftedAssignments(LOCAL_VALUE_DISPATCH);

    // The chain covers c0 and c1, so the lift produces two flattened
    // assignments. If this is 0 the pass stopped recognising the shape and the
    // rest of the test would pass vacuously.
    expect(pairs.map(p => p.prop)).toEqual(['c0', 'c1']);

    for (const pair of pairs) {
      expect(pair.then.length,
        `then-arm for this.${pair.prop} is empty; stack lowering will pad it with OP_0`,
      ).toBeGreaterThan(0);
      expect(pair.else.length).toBeGreaterThan(0);
    }
  });

  it('does not drop the then-arm when the value is a parameter', () => {
    const { pairs } = liftedAssignments(PARAM_VALUE_DISPATCH);
    expect(pairs.map(p => p.prop)).toEqual(['c0', 'c1']);
    for (const pair of pairs) {
      expect(pair.then.length,
        `then-arm for this.${pair.prop} is empty`,
      ).toBeGreaterThan(0);
    }
  });

  it('never emits a zeroed matched arm for a non-zero assignment', () => {
    for (const source of [LOCAL_VALUE_DISPATCH, PARAM_VALUE_DISPATCH]) {
      const result = compile(source);
      expect(result.success).toBe(true);
      expect(countByteSequence(result.artifact!.script, ZEROED_ARM)).toBe(0);
    }
  });

  it('CONTROL: an arm that really assigns 0n still pushes zero', () => {
    const result = compile(LITERAL_ZERO_DISPATCH);
    expect(result.success).toBe(true);
    // Two properties, two arms that legitimately push OP_0 against an OP_DUP
    // of the old value.
    expect(countByteSequence(result.artifact!.script, ZEROED_ARM)).toBe(2);
  });

  it('CONTROL: the in-arm (TicTacToe) shape still lifts and is unaffected', () => {
    const { pairs } = liftedAssignments(IN_ARM_VALUE_DISPATCH);
    expect(pairs.map(p => p.prop)).toEqual(['c0', 'c1']);
    for (const pair of pairs) {
      // A single load_prop of `turn` — the value already lands inside the arm,
      // so the fix must add nothing here.
      expect(pair.then.length).toBe(1);
    }
  });
});
